### DISCUSSION TITLE
How to load MicroPython on Seeed Arch Mix without using any debug probes

### DISCUSSION THREAD

**Discussion by skysurf76** [micropython/micropython#10439] [m.iMX RT / Teensy 4.x]
The following method is based on the information at https://wiki.seeedstudio.com/Arch_Mix/#flashing-arduino-bootloader-to-arch-mix

This was done on a Windows 10 computer.

I figured if they could use that NXP-MCUBootUtility to load the Arduino bootloader, maybe it could be used to load MicroPython firmware as well.  I used the following procedure...

-Download this zip file from GitHub at the following link...   https://github.com/JayHeng/NXP-MCUBootUtility/archive/refs/heads/master.zip
-Extract this zip file
-Download MicroPython firmware for the Arch Mix from https://micropython.org/download/SEEED_ARCH_MIX/
-Using a DuPont jumper wire connect 3.3 and RX on the Arch Mix board (pins 19 and 22).  This will activate the factory burned bootloader.
-Set the 4 pin dip switch on the Arch Mix board to 0001 (down down down up)
-In the directory of  the extracted zip file go to the bin directory and open NXP-MCUBootUtility.exe
-Plug in the Arch Mix board to your computer using the USB OTG connector.
-On the top left of the MCUBootUtility verify that next to "MCU" you have "i.MXRT105x" selected.
-Again in the top left of the MCUBootUtility GUI click "Boot Device Configuration".  In the drop down menu on the screen that comes up make sure "ISSI_IS25LPxxxA_IS25WPxxxA" is selected. Click "OK".
-On the left side of the MCUBootUtility GUI click "Connect to ROM".  If everything is successful you should see a blue dot above where you just clicked, and that button should now say "Reset Device".
-This is where we deviate from the Seeed Studio guide.  At the top of the MCUBootUtility GUI, find the tab labeled "Boot Device Memory" and select it.
-In the "Boot Device Memory" window, in the box next to "Start/Offset" put in 0x60000000.  I got the idea to try this value from https://micropython.org/download/SEEED_ARCH_MIX/
-In the box next to "bin/s19/hex:" click browse and select the MicroPython firmware you downloaded earlier.
-Now click the "Write(AutoErase)" button.  If everything is successful the bar at the bottom of the window will fill with green, and the firmware is flashed.
-You can now close the MCUBootUtility, unplug the Arch Mix board, remove the jumper wire between pins 19 and 22, and set the 4 switches back to 0010 (down down up down).
-Plug the Arch Mix back into your computer using the USB OTG connection, start Thonny with it set to detect port automatically, and you should get a REPL.  Or go into device manager and find the port and connect using whatever you wish.

MicroPython v1.19.1 on 2022-06-18; Seeed ARCH MIX with MIMXRT1052DVL5B

Type "help()" for more information.

>>> import gc
>>> gc.collect()
>>> gc.mem_free()
32779376
>>> 

Look at all that memory! :)

**Comment by robert-hh** [comment]
Thank you for detailing the firmware loading method and pointing to the NXP MCU Boot Utility. That is very convenient for devices without an embedded bootloader. The only drawback is, that it is only available for Windows hosts. I did not check whether it can be compiled from the sources for Linux or Max OS X.
@alphafred has made already and PR for a generic bootloader for the MIMXRT port, which makes firmware loading easy once a firmware image has been installed the first time. But that PR is still not merged.

**Comment by steeley** [comment]
Setting those DIP switches and jumper just cause the board to have no USB port so it is impossible to connect to it with the utility. Are you sure these are the right settings? I've often found errors in Seeed docs....

**Reply by robert-hh** [reply]
The DIP switch change & jumper cause the board to boot from the serial interface, allowing to upload a special bootstrap loader for firmware update to RAM. Once that is done, you have to set the switches back, remove the jumper and push reset. Then the device boots again from flash, and the loaded firmware should start.

**Reply by skysurf76** [reply]
Steeley, it's been awhile since I wrote that guide.  It is always possible it has an error in it.  I will re-verify the procedure this evening by re-flashing one of my boards to make sure the steps are correct.  If you get it working in the mean time let us know.

As robert-hh said, with the jumper installed between the pins and the DIP switches set correctly the board will boot into a state where a standard serial port is not present.  The NXP-MCUBootUtility will be able to upload the new firmware to the board when it is in this state though.  I don't have a deep understanding of how this stage works I just know that the NXP-MCUBootUtility is able to communicate with the board.  After the utility uploads the new firmware do what robert-hh said as far as removing the jumper and resetting the DIP switches and it should boot from the firmware you just installed.

**Comment by steeley** [comment]
thanks to all for the quick reply - looks like I had a USB port issue. I tried it on a different computer and the utility found the USB device
( appears as some kind of weird HID that the OS ignores)
That utility is very clunky though ... easy to brick a device I'm sure!

**Reply by robert-hh** [reply]
You cannot brick a MIMXRT device, because the initial loader is located in ROM. So you can always re-load a firmware.

**Reply by skysurf76** [reply]
Excellent, glad to hear you got it working steeley.  I couldn't remember what showed up in device manager when you have the jumper on and the DIP switches set.  Whatever it appears as to the OS, apparently an HID device,  the NXP-MCUBootUtility has the ability to talk to it.

And yes, that utility is definitely a bit clunky, however it works, so its good enough. :)

**Reply by steeley** [reply]
yes, Rom Bootloader definitely a plus..
thanks

**Comment by steeley** [comment]
latest firmware seems to be loaded now and I have a REPL in Thonny.
Any idea what the built in RGB LED pins are? I thought it was 52 53 54 as it appears to be analog RGB.
in fact the whole pin addressing on this board is a bit odd 
This board originally had Arduino on it and that was working ok.

**Reply by robert-hh** [reply]
According to pins.csv, it is "LED_GREEN", "LED_RED", "LED_BLUE", or 59, 60, 61, or Pin.board.LED_GREEN, .Pin.board.LED_RED, Pin.board.LED_BLUE, or the respective GPIO numbers as string or Pin.cpu.GPIO....

**Reply by steeley** [reply]
thanks that seems to work.
Where is this 'pins.csv' file? I've been unable to locate it.

**Reply by robert-hh** [reply]
It's here: https://github.com/micropython/micropython/tree/master/ports/mimxrt/boards/SEEED_ARCH_MIX
You may as well look into the documentation for the assignment of devices like UART, SPI, I2C, PWM to pins.
https://docs.micropython.org/en/latest/mimxrt/pinout.html

**Reply by steeley** [reply]
Ahh thanks... I had a spreadsheet of pins from Seeed but that just relates names to schematic - not MicroPython or Arduino names.

Just tested and it works but the logic is wrong - ledRed.off() (or Green or Blue) turns the led(s) on. Found the schematic and confirmed leds are wired with the anodes to common to v+ then MCU pins pull to ground to turn on.

**Reply by robert-hh** [reply]
> Just tested and it works but the logic is wrong - ledRed.off() 

That kind of confusion is common to all ports. The on() and off() methods just set the level high or low. Personally I do not like the on() & off() methods. A pin cannot be on or off. You may have a look at the signal class, which allows to specify, whether a level for a signal is straight or inverted.

**Reply by steeley** [reply]
Yes a bit of an issue. I do a lot more desktop software(and analog hardware) than embedded stuff, but for a method like xxx.on() xxx.off() applied to a specific piece of fixed hardware,
I always make that mean "device active"=  xxx.on() and "device inactive" = xxx.off(). Agreed a pin being 'on' or 'off' is perhaps not technically correct. Signal class looks useful.

thanks again.
